src/megatron/bridge/models/conversion/model_bridge.py

```diff
@@ -103,68 +102,50 @@ class WeightConversionTask(Generic[MappingT]):
 
-        Args:
-            megatron_weights: Megatron weight tensor to convert (may be None if not owned by this rank).
-            megatron_module: Megatron module that owns the parameter (may be None).
-
-        Returns:
-            Dictionary mapping HuggingFace parameter names to converted tensors.
-
-        Raises:
-            ValueError: If required fields for Megatron->HF conversion are missing.
-        """
-
-        return self.mapping.megatron_to_hf(megatron_weights, megatron_module, self.param_name)
-
-
-def _adjust_layer_number_to_global(name: str, layer_offset: int) -> str:
-    """Adjust layer number from local to global numbering."""
-    if "layers." not in name:
-        return name
-
-    local_layer_number = int(name.split("layers.")[1].split(".")[0])
-    global_layer_number = local_layer_number + layer_offset
-    name = name.replace(
-        f"layers.{local_layer_number}.",
-        f"layers.{global_layer_number}.",
-    )
+def _megatron_local_name_to_global(
+    models: MegatronModule | List[MegatronModule],
+    config: TransformerConfig,
+    param_name: str,
+    vp_stage: Optional[int] = None,
+) -> str:
+    """Adjust layer number and expert number from local to global numbering."""
+    # PP
+    pp_group = parallel_state.get_pipeline_model_parallel_group()
+    if "layers." in param_name and get_pg_size(pp_group) > 1:
+        match = re.match(r"^(.+?\.layers\.\d+)", param_name)
+        assert match is not None
+        layer_prefix = match.group(1)
+        _, layer_module = get_module_and_param_from_name(models=models, param_name=layer_prefix, vp_stage=vp_stage)
+
+        local_layer_number = int(param_name.split("layers.")[1].split(".")[0])
+        global_layer_number = layer_module.layer_number - 1
+        param_name = param_name.replace(
+            f"layers.{local_layer_number}.",
+            f"layers.{global_layer_number}.",
+        )
 
-    return name
+    # EP
+    ep_group = parallel_state.get_expert_model_parallel_group()
+    if ".mlp.experts.linear_fc" in param_name and get_pg_size(ep_group) > 1:
+        num_experts = config.num_moe_experts
+        num_experts_per_rank = num_experts // ep_group.size()
+        # Handle weight and bias parameters separately
+        if ".weight" in param_name:
+            local_expert_number = int(param_name.split(".weight")[-1])
+            global_expert_number = num_experts_per_rank * ep_group.rank() + local_expert_number
+            param_name = param_name.replace(
+                f".weight{local_expert_number}",
+                f".weight{global_expert_number}",
+            )
+        elif ".bias" in param_name:
+            local_expert_number = int(param_name.split(".bias")[-1])
+            global_expert_number = num_experts_per_rank * ep_group.rank() + local_expert_number
+            param_name = param_name.replace(
+                f".bias{local_expert_number}",
+                f".bias{global_expert_number}",
+            )
```

Comment:
nit: code is duplicated except for the weight vs bias key. make it into a function to share the logic?